MFA in the new authorization server

matrix-amiel opened the pull request:
Hi @sjpadgett 

Here I add a shortcode that implements MFA for Grant password (custom) API.
The process is simple, In the case that the user already registered to MFA. the response of `openemr/oauth2/default/token` with `grant_type=password` will be -
```
HTTP/1.1 403 Forbidden
{
    "error": "mfa_required",
    "error_description": "MFA required, The authorization server expects to `mfa_token` parameter in the request body.",
    "message": "MFA required, The authorization server expects to `mfa_token` parameter in the request body."
}
```
When the user sends the valid MFA code in addition to existing parameters, the login will be finished successfully
```
HTTP/1.1 200 OK
{
    "id_token": "eyJ0eXAi..................JsnyqyMKLDf1UEnrc_ipV0wR4oe8NoVSDZtPeM5Ane65tw_V-xuiaK3A1p2ZIN2dLLjhWlgqsrw1_3A",
    "token_type": "Bearer",
    "expires_in": 3599,
    "access_token": "eyJ0eXAiOiJKV1QiLC...........wF2BTI1ehDHDPYa4yY2lSwV89wfHXvc1X8vAbuDWE-f9c8wNhDpJwg7PbRJA",
    "refresh_token": "def50200d27c5df.................115d"
}
``` 

Thanks
Amiel

sjpadgett commented on src/Common/Auth/OpenIDConnect/Entities/UserEntity.php:
```diff
@@ -90,6 +92,30 @@ protected function getAccountByPassword($userrole, $username, $password, $email
                 $this->setIdentifier(UuidRegistry::uuidToString($uuid));
+
+                //check if MFA required
+                $mfa = new MfaUtils($id);
+                if ($mfa->isMfaRequired() && is_null($mfaToken)) {
```
I'm thinking you could grab the token here either by adding a method to your MfaUtils() i.e isMfaToken() then return token/null, grabbing the token from the request in your utility or grab the token from request here.

sjpadgett replied:
Also will make it easier to disengage this flow from password grant.

sjpadgett reviewed (changes requested):
Hi @matrix-amiel 
The way you have submitted PR is fine where I just made a couple suggestions you're welcome to follow or not.
My concern is by using the password grant to enforce the third party authorization, we miss the chance to turn this into an authorization code grant flow. Even though it's an openemr feature, to the identity server, the way openemr MFA should be treated is as a third party provider.

So in that regard, the flow should start as an auth code grant request bypassing normal auth server sign in(the hook is preliminarily supported in the `oauthAuthorizationFlow()` method extending the `if ($this->providerForm)`  to bypass) then dealing with the identity as a normal auth grant flow.

Eventually we need to remove ourselves from users logging in with a username/password and using our new identity server to handle a more secure code grant flow plus providing a way for vendors to use third party authorization if they prefer(essentially what you desire here).

You can however thank Brady for us even supporting legacy. Personally I wanted to force the issue now so vendor clients would get on board with the more secure methods.

I sincerely hope i've been able to communicate my concerns understandably. If not, please comment and thanks for the PR.

sjpadgett commented:
@adunsulag this may of note to you.

bradymiller commented:
hi @matrix-amiel ,
To clarify, when passing the mfa_token parameter:
1. For TOTP method, is it the 6 digit code?
2. For U2F method, what is it? Is it a token captured on the client side that collects if from the U2F device? If so, is this token changing (like TOTP), or is it static for the U2F device?

matrix-amiel commented:
Hi @sjpadgett 
I definitely understand your concerns, and I agree that the right way is using only the code grant method. for now, we are using user/password because we don't have time to make changes in the login application styling that presents to customers in these days. We plan to move to code grant, even with third parties in the next steps.

However, I ready to implement 2FA in the auth code grant also, there is an OpenID standard for the API calls? I didn't find it... if you have some advice on where to add it to code, I'd appreciate you.  
Hope to try it tomorrow.

matrix-amiel commented:
@bradymiller 
Hi!
1. yes. TOTP is 6 6 digit code.
2. I'll be honest, I don't have such a USB device and I don't have any way to test it :grimacing: I just copied the code from main_screen.php, if you have someone that can help with this info and testing it will be great....

sjpadgett commented:
Hi @matrix-amiel (Note: I understand what you're trying to do but i'm overstating in the following to perhaps educate others seeking to support the public app/client)

The only 2F flow that possibly would apply would be the implicit flow. Here again your're face with some of the concerns as the password flow. The following excerpt from OAuth 2.0 explains this pretty well:

The Implicit flow was a simplified OAuth flow previously recommended for native apps and JavaScript apps where the access token was returned immediately without an extra authorization code exchange step.
It is not recommended to use the implicit flow (and some servers prohibit this flow entirely) due to the inherent risks of returning access tokens in an HTTP redirect without any confirmation that it has been received by the client.
Public clients such as native apps and JavaScript apps should now use the authorization code flow with the PKCE extension instead.

This is me now: I support the PKCE extension which is a code challenge.  This version of code flow is meant for public apps that can't or rather shouldn't store a secret.

However, appears to me that you want to be a confidential app where truely the MFA has already authenticated user and authorized access permissions for openemr api access i.e probably one of our users more so than patient access, and not do the first leg(authorization code exchange) whereby bypassing the authorization and identity server capabilities to get to a token the resource/dispatcher server understands(Phew, that was hard to get out!). Understandable and you could use the implicit flow hitting the default/authorize endpoint.

Depending on how much energy I have left today/tonight i'll do one of two things. Go ahead and add the implicit flow to the server where i'll leave to you to finish hooking in bypassing user session. Or post here where you need to add the flow in source.
Really easier for me to do quickly.

@bradymiller  and @adunsulag You guys keep me from my sometimes single mindedness if needed!;)

sjpadgett commented:
Looking into this further, the implicit and client flow will/should not return a refresh token making the resource/password grant appropriate for now. Otherwise, client will have to request a new access token for each api request. Which when you think about it, is no different to a new access token request via a refresh token. umm, bringing us full circle to my notion that in many cases, why a refresh token!
So, until I can get the entire third party authorization in place we'll go with this PR for now.
Let me know if finished and I'll bring in or you can do the flow by adding a new getAuthorizationServer() init params supporting the implicit flow adding a new method to handle the endpoint request e.g oauthImplicitFlow().
Follow https://oauth2.thephpleague.com/authorization-server/implicit-grant/ for setup however, I wonder if the method will allow adding the $refreshTokenRepository? 
I could set this up but would take a couple days to get to. Up to you Amiel.

matrix-amiel commented:
@sjpadgett 
Sorry but I don't understand the connection between 2FA to implicit flow.
implicit flow sounds for me very not secure and does not match to medical record application.
I'll try to complete the PR today and to push 2FA for the Authorization code grant also

matrix-amiel commented:
Hi

I added MFA to the screens of authorization server login:
![image](https://user-images.githubusercontent.com/17809866/101340403-594f2980-3888-11eb-8120-9f383d8fa207.png)
I don't add the U2F method, I don't have the device to check it and I don't want to create a js error. for now, it's working for TOTP method that is a very comfortable way to use 2FA.

Regardless of my changes, I found bugs (that weren't found when I checked it before a few weeks), the CSFR token is not verified in the request, and the sessions that created in the oauthAuthorizationFlow not found in the userLogin method, is it working for you? what I did wrong?

Hope now the work is more complete
tell me how to go ahead...

Thakns

sjpadgett commented:
Hi @matrix-amiel 
Sorry, I thought we were talking about 2 legged not 2 factor!

I'm retesting code flow now

sjpadgett commented:
@matrix-amiel The current master code is working okay all the way from new tokens through a fhir patient get.

When speaking about CSRF are you seeing an error with the forms CSRF or the tokens state attribute?

matrix-amiel commented:
The code grant process not working for me...
1.I call to `openemr/oauth2/default/authorize` using GET http, in the \OpenEMR\RestControllers\AuthorizationController::oauthAuthorizationFlow created a few SESSION parameters (csrf,scopes,client_id,csrf_private_key)
2. the AuthorizationController::oauthAuthorizationFlow redirects the call to `openemr/oauth2/default//provider/login`.
3.  when the process initiated in the `openemr/oauth2/authorize.php` file seems the scripts need to know what was stored in the SESSIONS in the previous call, but the data not available and all the sessions are empty. 

What I'm missing...

Regarding the PR do you see any changes I need to do?

Thank you

sjpadgett commented:
The session should get restored in oauthAuthorizeToken() if you send the correct request back after the code and user are validated in  authorizeUser() where user $session_cache is stored for the last request in flow.
Make sure you have the grant_type set to authorization_code in final request. Hopefully not sending password grant.

PR looks okay for me. Just ensure session is populated after dialog and repopulates in oauthAuthorizeToken()

matrix-amiel commented:
Sorry, I'm working on that a lot of hours - 
I have no idea why -
```
 public static function oauthSessionStart($web_root): void
    {
        session_start([
            'cookie_samesite' => "Lax",
            'cookie_secure' => true,
            'name' => 'authserverOpenEMR',
            'cookie_httponly' => self::$use_cookie_httponly,
            'cookie_path' => ((!empty($web_root)) ? $web_root . '/oauth2/' : '/oauth2/'),
            'gc_maxlifetime' => self::$gc_maxlifetime,
            'sid_bits_per_character' => self::$sid_bits_per_character,
            'sid_length' => self::$sid_length,
            'use_strict_mode' => self::$use_strict_mode,
            'use_cookies' => self::$use_cookies,
            'use_only_cookies' => self::$use_only_cookies
        ]);
    }
```
can not create a cookie in the browser, seems that the sessions lost after the redirect from `oauthAuthorizationFlow` to login action.

Would happy to any hint... need to get some rest from this... 
(I didn't mean to deep into world of sessions :pensive: )

sjpadgett commented:
> 'cookie_samesite' => "Lax",

Try changing to None. If necessary i'll pull in and test

matrix-amiel commented:
> > 'cookie_samesite' => "Lax",
> 
> Try changing to None. If necessary i'll pull in and test
> 
> setting up and testing now.

Thanks... :pray:
Still not working for me.. let me know if works for you in regular mode (without MFA) or  I ruined something...

sjpadgett commented:
Yep, it broke me! I'll figure it out and get back to ya. Kind of strange how it's breaking normal flow....

bradymiller commented:
line 34 in login.php looks odd (at work and was able to check out the code
on my cell)

On Tue, Dec 8, 2020 at 12:36 PM Jerry Padgett <notifications@github.com>
wrote:

> Yep, it broke me! I'll figure it out and get back to ya. Kind of strange
> how it's breaking normal flow....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/4086#issuecomment-740990929>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEDODQLUDDNFJ5UMK23RLST2E5DANCNFSM4UPGZNLQ>
> .
>

sjpadgett commented:
@yep but just the header to the dialog so won't hurt

bradymiller commented:
Agreed that is a minor issue.

I may by missing something, but I don't see how U2F will work if try to separate logic from the javascript client component.

sjpadgett commented:
@matrix-amiel 
My initial issue was my environment/keys got messed up somehow. Need to look into that. However, this PR works for me from my oeHttp client. I'll test with another client in am...

matrix-amiel commented:
> I may by missing something, but I don't see how U2F will work if try to separate its authentication logic from the javascript client component.

Every client will need to write the javascript code in his client app in this case. do you recommend to remove U2F authentication from API login? (again I don't have a device to test it)

bradymiller commented:
Quick question @matrix-amiel ,
If the login.php (code grant process) solution covers both TOTP/U2F (definitely can do since would be just like done in interface/main/main_screen.php) would there still be a need for MFA support in password grant flow?

matrix-amiel commented:
Yes, we are using password grant for now and needed MFA using TOTP... this was the start point of my PR...

bradymiller commented:
The problem here is that U2F requires a successful install then followed by a challenge cascade:
https://developers.yubico.com/U2F/Libraries/Using_a_library.html

So I don't think it can be done via the password_grant flow (unless I am missing something). So, this strategy on the password grant flow will cause inconsistency (it will never work for users that are set to use U2F and will also not likely work for users that are set up to both support TOTP and U2F).

bradymiller commented:
If I am correct that U2F will be impossible to support via password grant, then would the following strategy work for you @matrix-amiel for the password grant  :
1. If the password grant includes the mfa_token, then check do TOTP (and fail if TOTP not set for user)
2. If the password grant does not include the mfa_token, then just do normal auth

matrix-amiel commented:
@bradymiller 
To be sure, in password grant in the case that the user sets TOTP and the mfa_token not sent in the parameters, I need to return 403, right? (if he has only U2F and he doesn't send mfa_token I continue with normal auth)

bradymiller commented:
@matrix-amiel ,
For the password grant flow I don't think there are really any rules :) (since it is considered not secure and not recommended to use). 
Just seems inconsistent to force TOTP but skip U2F.
And seems more consistent to just support TOTP only if the mfa_token is sent.
This then places the decision on the client app whether wish to use MFA or not for the password grant.

Now the code grant process is another matter and the plan there is to force mfa if this is set for the user (just like a standard login is done).

sjpadgett commented:
Hi @bradymiller and @matrix-amiel 
I tested this against a client that also supports token nonce, state and token signature with success.
Brady, I get your point concerning client responsibility to let server know the request is a third party authorized request. This could be setup as a flag at beginning of flow where the providerform flag is set to inform server login to grab the pin instead of a check every sign in.
I guest if a user is registered with a mfa and client wanted to bypass, client would just use the password grant thereby bypassing mfa auth.

Amiel, I would think a 403 status would be fine or a 400. Most clients will take note of a 403 or 400 where a 401 generally means an invalid token in our case.

I'm trying to understand what the password grant has to do with this as all the client is doing is passing in the result of the third party users authentication with a stored password/username of the first leg of code flow so server can authorize via a pin. The password flow hasn't anything to do here imo.

Still I like this and didn't realize how well done the OpenEMR MFA is. I would like to see the decision if a third party auth at beginning of flow where the providerform is set if possible.

matrix-amiel commented:
According to the last @bradymiller comment, I removed U2F check from password grant and added U2F authentication to the code grant process (tested as I can without device)

sjpadgett commented on oauth2/provider/login.php:
```diff
@@ -31,8 +32,10 @@
+                    <?php } elseif (empty($mfaRequired) && $mfaRequired && $mfaType === MfaUtils::TOTP) { ?>
                         <h4 class="card-title mb-4 mt-1"><?php echo xlt('TOTP Verification'); ?></h4>
+                    <?php } elseif (empty($mfaRequired) && $mfaRequired && $mfaType === MfaUtils::U2F) { ?>
+                        <h4 class="card-title mb-4 mt-1"><?php echo xlt('U2F Key Verification'); ?></h4>
```
empty($mfaRequired) && $mfaRequired will always be false won't it. Also, not PHP8 compatible e.g  empty($mfaRequired) && !empty($mfaRequired) or empty($mfaRequired) && $mfaRequired === ?.

matrix-amiel replied:
Done

sjpadgett commented on oauth2/provider/login.php:
```diff
@@ -66,11 +69,32 @@
                         <?php } ?>
 
-                        <?php if (empty($authorize) && $mfaRequired && $mfaType === 'TOTP') { ?>
+                        <?php if (empty($authorize) && $mfaRequired ) { ?>
```
again I think PHP wants explicite conditionals from what I've seen Brady fix many in core. e.q !empty($mfaRequired) I could be wrong here as I haven't researched.

sjpadgett commented on oauth2/provider/login.php:
```diff
@@ -66,11 +69,32 @@
-                        <?php if (empty($authorize) && $mfaRequired && $mfaType === 'TOTP') { ?>
+                        <?php if (empty($authorize) && $mfaRequired ) { ?>
+
+                            <?php if ($mfaType === MfaUtils::TOTP) { ?>
```
Would it not be better to set any required variables in controller instead of calling method on form render?

matrix-amiel replied:
Do you mean to change MfaUtils::TOTP to regular variable that created in the AuthorizationController?

sjpadgett replied:
yep just to keep form consistent

matrix-amiel replied:
Done

sjpadgett commented on oauth2/provider/login.php:
```diff
@@ -66,11 +69,32 @@
                             <h5><?php echo xlt('Provide TOTP code') ?></h5>
                             <div class="form-group">
-                                <input class="form-control" autocomplete="false" placeholder="<?php echo xla("Enter the code"); ?>" type="text" name="mfa_token">
+                                <input class="form-control" autocomplete="false" placeholder="<?php echo xlt("Enter the code"); ?>" type="text" name="mfa_token">
```
Nitpicking here but  xlt("Enter required authentication code")

matrix-amiel replied:
Done

sjpadgett commented:
again I don't get what the password grant has to do with this. Ya can't do a 2F in the password flow anyway so I don't see why the client cares if user is also registered as a mfa!
I'm prob not communicating very well and I tend to go alone with the dev if it works for them and it doesn't break features intent but, i'll probably change the workflow some later to be more consistent with a third party auth workflow.

I do like that this got done. Pretty cool stuff and I don't have a dongle to test with but, now i've played with MFA, i'll get one so I can test in future.

bradymiller commented:
Regarding password grant, in an ideal world we would not even bother with
mfa. But matrix has a specific use case to support totp. If a mfa_token
parameter is passed on the password grant, then the following should take
place:
1. Check if user has totp set up (if not, then fail auth)
2. If user has totp set up, then do the check

The key here is we let the application decide whether to use totp. If it
includes the mfa_token on the password grant call, then do above. If do not
include mfa_token on the password grant call, then skip mfa (even if the
user is set up for it).

On Wed, Dec 9, 2020 at 8:37 AM Jerry Padgett <notifications@github.com>
wrote:

> again I don't get what the password grant has to do with this. Ya can't do
> a 2F in the password flow anyway so I don't see why the client cares if
> user is also registered as a mfa!
> I'm prob not communicating very well and I tend to go alone with the dev
> if it works for them and it doesn't break features intent but, i'll
> probably change the workflow some later to be more consistent with a third
> party auth workflow.
>
> I do like that this got done. Pretty cool stuff and I don't have a dongle
> to test with but, now i've played with MFA, i'll get one so I can test in
> future.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/4086#issuecomment-741892615>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEDOAQ4C7VGIRUXODX5MTST6RVNANCNFSM4UPGZNLQ>
> .
>

bradymiller commented:
btw, the U2F option in MFA is really cool (got a really cheap U2F awhile
back to test it and can test this out when have some time; will probably be
after this is in the codebase though).

Amiel, another fun thing to consider on the code grant process is the user
can be set up for both TOTP and U2F at the same time :)

On Wed, Dec 9, 2020 at 8:51 AM Brady Miller <brady.g.miller@gmail.com>
wrote:

> Regarding password grant, in an ideal world we would not even bother with
> mfa. But matrix has a specific use case to support totp. If a mfa_token
> parameter is passed on the password grant, then the following should take
> place:
> 1. Check if user has totp set up (if not, then fail auth)
> 2. If user has totp set up, then do the check
>
> The key here is we let the application decide whether to use totp. If it
> includes the mfa_token on the password grant call, then do above. If do not
> include mfa_token on the password grant call, then skip mfa (even if the
> user is set up for it).
>
> On Wed, Dec 9, 2020 at 8:37 AM Jerry Padgett <notifications@github.com>
> wrote:
>
>> again I don't get what the password grant has to do with this. Ya can't
>> do a 2F in the password flow anyway so I don't see why the client cares if
>> user is also registered as a mfa!
>> I'm prob not communicating very well and I tend to go alone with the dev
>> if it works for them and it doesn't break features intent but, i'll
>> probably change the workflow some later to be more consistent with a third
>> party auth workflow.
>>
>> I do like that this got done. Pretty cool stuff and I don't have a dongle
>> to test with but, now i've played with MFA, i'll get one so I can test in
>> future.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openemr/openemr/pull/4086#issuecomment-741892615>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AACEDOAQ4C7VGIRUXODX5MTST6RVNANCNFSM4UPGZNLQ>
>> .
>>
>

sjpadgett commented:
Okay just let me know @matrix-amiel and i'll bring in. I want to adjust a little.

matrix-amiel commented:
I haven't known it possible...:flushed:

Added support for this case...

sjpadgett commented:
I assume this is finished. @matrix-amiel anymore? I'll test again and i'm guessing Brady will test w/dongle once in codebase.
I'll fix psr on my next PR.

sjpadgett commented:
Merging so I can get on the Client grant and JWK issuance for client registration. Plus I have a SMART PR pending which this will create conflicts.
I'll test, fix and notify you of any issues I come across regarding this @matrix-amiel. 
Thanks for this as we always appreciate Matrix contributions.
